Migrate hooks documentation from wiki (#2782)
jenkins-infra/jenkins.io: content/doc/book/managing/groovy-hook-scripts.adoc

:author:
:email: jenkinsci-docs@googlegroups.com
:sectanchors:
:toc: left
endif::[]

= Groovy Hook Scripts

In several places inside Jenkins, a series of "hook scripts" get executed to allow some actions to take place in reaction to some key
events.

These scripts are written in Groovy, and get executed inside the same JVM as Jenkins, allowing full access to the domain model of Jenkins. 
For given hook _HOOK_, the following locations are searched:

* `+WEB-INF/HOOK.groovy+` in `+jenkins.war+`
* `+WEB-INF/HOOK.groovy.d/*.groovy+` in the lexical order in `+jenkins.war+`
* `+$JENKINS_HOME/HOOK.groovy+`
* `+$JENKINS_HOME/HOOK.groovy.d/*.groovy+` in the lexical order

`+HOOK.groovy.d+` is suitable to avoid conflicts — multiple entities can
insert stuff into the hook without worrying about overwriting each
other's code.

The following events use this mechanism:

* *init*: Post-initialization script
* *boot-failure*: Boot failure hook

== Post initialization script (init hook)
You can create a Groovy script file `+$JENKINS_HOME/init.groovy+`, or
any `+.groovy+` file in the directory `+$JENKINS_HOME/init.groovy.d/+`,
to run some additional things right after Jenkins starts up. 

> The script is executed happen at the very end of the initialization, and therefore this can be used to pre-configure Jenkins for a particular OEM situation.

Reviewer: ```suggestion
The groovy scripts are executed at the end of Jenkins initialization.
```
